prime-router/src/main/kotlin/azure/ActionHistory.kt

```diff
@@ -0,0 +1,336 @@
+        }
+
+        val reportFile = ReportFile()
+        reportFile.reportId = report.id
+        reportFile.nextAction = TaskAction.none
+        if (report.sources.size != 1) {
+            error("An external incoming report should have only one source.   Report ${report.id} had ${report.sources.size} sources")
+        }
+        val source = (report.sources[0] as ClientSource)
+        reportFile.sendingOrg = source.organization
+        reportFile.sendingOrgClient = source.client
+        reportFile.schemaName = report.schema.name
+        reportFile.schemaTopic = report.schema.topic
+        reportFile.bodyUrl = report.bodyURL
+        reportFile.bodyFormat = report.getBodyFormat().toString()
+        reportFile.itemCount = report.itemCount
+        reportsReceived[reportFile.reportId] = reportFile
+    }
+
+    /* Table structure here for reference during development. Might be out of date.
+        public ReportFile(
+            UUID           reportId,
+            Integer        actionId,
+            TaskAction     nextAction,
+            OffsetDateTime nextActionAt,
+            String         sendingOrg,
+            String         sendingOrgClient,
+            String         receivingOrg,
+            String         receivingOrgSvc,
+            String         schemaName,
+            String         schemaTopic,
+            String         bodyUrl,
+            String         bodyFormat,
+            byte[]         blobDigest,
+            Integer        itemCount,
+            OffsetDateTime wipedAt,
+            OffsetDateTime createdAt
+        */
+    /**
+     * Use this to record history info about an internally created report.
+     */
+    fun trackCreatedReport(
+        event: Event,
+        report: Report,
+        service: OrganizationService,
+        validatedRequest: ReportFunction.ValidatedRequest
+    ) {
+        if (reportAlreadyTracked(report.id)) {
+            error("Bug:  attempt to track history of a report ($report.id) we've already associated with this action")
+        }
+
+        val reportFile = ReportFile()
+        reportFile.reportId = report.id
+        reportFile.nextAction = event.action.toTaskAction()
+        reportFile.nextActionAt = event.at
+        reportFile.receivingOrg = service.organization.name
+        reportFile.receivingOrgSvc = service.name
+        reportFile.schemaName = report.schema.name
+        reportFile.schemaTopic = report.schema.topic
+        reportFile.bodyUrl = report.bodyURL
+        reportFile.bodyFormat = report.getBodyFormat().toString()
+        reportFile.itemCount = report.itemCount
+        reportsOut[reportFile.reportId] = reportFile
+    }
+
+    /**
+     * Save the history about this action and related reports
+     */
+    fun saveToDb(db: DatabaseAccess, txn: Configuration? = null) {
+        if (txn != null) {
+            insertAll(txn)
+        } else {
+            db.transact { innerTxn -> insertAll(innerTxn) }
+        }
```

Comment:
My vision is that `DatabaseAccess` will become very tiny - it will hold the generic stuff only.   So a second pass will go thru and simplify it.

All other table-specific code will go in other classes, like ActionHistory.  Sorta like DAOs but not that formal or fancy.